include/nanobind/eigen/dense.h

```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+template <typename T> struct type_caster<T, enable_if_t<is_eigen_sparse_v<T>>> {
+    using Scalar = typename T::Scalar;
+    using StorageIndex = std::remove_reference_t<decltype(*std::declval<T>().outerIndexPtr())>;
+    using Index = typename T::Index;
+    static constexpr bool rowMajor = T::IsRowMajor;
+
+    using ScalarCaster = make_caster<tensor<Scalar>>;
+    using StorageIndexCaster = make_caster<tensor<StorageIndex>>;
+    using IndexCaster = make_caster<Index>;
+
+    NB_TYPE_CASTER(T, const_name<rowMajor>("scipy.sparse.csr_matrix[",
+					   "scipy.sparse.csc_matrix[")
+		   + make_caster<Scalar>::Name + const_name("]"));
+
+    ScalarCaster scalar_caster;
+    StorageIndexCaster storageindex_caster;
+    IndexCaster index_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+	if (!src) {
+	    return false;
+	}
+
+	auto obj = borrow<object>(src);
+	object sparse_module = module_::import_("scipy.sparse");
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
+
+	if (!obj.type().is(matrix_type)) {
+            try {
+                obj = matrix_type(obj);
+            } catch (const python_error &) {
+                return false;
+            }
+        }
+
+	if (!scalar_caster.from_python(obj.attr("data"), flags, cleanup)) {
+	    return false;
+	}
+        auto values = scalar_caster.value;
+
+	if (!storageindex_caster.from_python(obj.attr("indices"), flags, cleanup)) {
+	    return false;
+	}
+        auto innerIndices = storageindex_caster.value;
+
+	if (!scalar_caster.from_python(obj.attr("indptr"), flags, cleanup)) {
+	    return false;
+	}
+        auto outerIndices = storageindex_caster.value;
+
+	// FIXME: Write with caster.from_python
+	auto shape = cast<std::tuple<Index, Index>>(obj.attr("shape"));
+
+	if (!index_caster.from_python(obj.attr("nnz"), flags, cleanup)) {
+	    return false;
+	}
+        auto nnz = index_caster.value;
+
+        value = EigenMapSparseMatrix<Scalar,
+                                     T::Flags &(Eigen::RowMajor | Eigen::ColMajor),
+                                     StorageIndex>(std::get<0>(shape),
+						   std::get<1>(shape),
+                                                   std::move(nnz),
+                                                   outerIndices.data(),
+                                                   innerIndices.data(),
+                                                   values.data());
+
+        return true;
+    }
+
+    static handle from_cpp(T &&v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        if (policy == rv_policy::automatic ||
+            policy == rv_policy::automatic_reference)
+            policy = rv_policy::move;
+
+        return from_cpp((const T &) v, policy, cleanup);
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+	// FIXME: Looks like UB. Is there anything that can be done about this?
```

Comment:
did pybind11 use the same code?

Reply:
Yes. https://github.com/pybind/pybind11/blob/master/include/pybind11/eigen/matrix.h#L677